app_dart/lib/src/service/scheduler.dart

```diff
@@ -231,7 +231,9 @@ class Scheduler {
       retryOptions: retryOptions,
     );
     final YamlMap configYaml = loadYaml(configContent) as YamlMap;
-    return schedulerConfigFromYaml(configYaml).writeToBuffer();
+    pb.SchedulerConfig schedulerConfig =
+        await schedulerConfigFromYaml(configYaml, ensureBringupTargets: true, slug: commit.slug);
```

Comment:
Instead of passing `RepositorySlug`, we should actually have this validate against `CiYaml` which contains this protobuf. `CiYaml` already contains the extra state necessary to do this validation, like repository slug and branch.

Reply:
Umm maybe my understanding is not correct, I think the most common way to obtain a CiYaml class is to call the scheduler.getCiYaml(commit) function in scheduler.dart. But I think `getCiYaml` function calls on the underlying `_downloadCiYaml` function, which relies on `githubFileContent` function to generate a scheduler config. My understanding is that the dependency goes like `getCiYaml` --depends on--> `_downloadCiYaml` --depends on--> `githubFileContent`. where `getCiYaml` produces CiYaml and `_downloadCiYaml` produces schedulerConfig.

Maybe I am wrong, but if we are supplying CiYaml class in `_downloadCiYaml`, I thought we are using CiYaml class which is produced by scheduler config which is eventually produced by the function itself. And therefore we have a dependency loop.

Reply:
One way would be to move this `schedulerConfigFromYaml` directly into `CiYaml` as a named constructor. `CiYaml` is a relatively new class, so that's why it wasn't in there in the first place. You can still keep the http requests outside of `CiYaml` by having `CiYaml.fromYaml(YamlMap yaml);`.

For the cache section, you can still access the underlying proto via `ciYaml.value`

Reply:
Thanks for the advice! I experimented around and umm my current plan is to move the entire logic of getCiYaml and _downloadCiYaml into under CiYaml class. I figured that I would have a new `.fromYaml` function in CiYaml which is a wrapper of githubFileContent and takes in parameters such a slug and branch. I am not entirely sure about the cache section, my understanding is that we create a `.value` function in CiYaml which wraps the caching logic from getCiYaml, and therefore move everything from getCiYaml and _downloadCiYaml to CiYaml class as well. 

For schedulerConfigFromYaml as a named constructor inside CiYaml, I am not sure how to supply a CiYaml parameter from within CiYaml. My current plan is to still use slug information, and supply it with the slug and branch information obtained from within the CiYaml class. And for other classes depending on just the schedulerConfig portion of the CiYaml class, I am planning to provide a CiYaml class with every field set to null except the schedulerConfig part. Maybe I misunderstood, does it sound correct to move everything into CiYaml, and to still supply branch and slug information to these functions, after they have become CiYaml class functions? Thank you!

Reply:
We want to minimize deps in models (that ways its easier to share the models in other code bases), so avoid adding the http work to `CiYaml` The caching logic will remain in `Scheduler`.

I don't understand your second point. Can you clarify, or show a snippet of the issue you're running into?